### PR TITLE
feat: Perform asciicast conversion sanity checks

### PR DESCRIPTION
**Pull request by elimt**
Check that the conversion to asciicast is done on `exec` or `shell` session program. 
Unsupported session programs will throw an error.

### CODE REVIEW

File: internal/bsr/convert/convert.go [hunk 1]
```diff
@@ -55,20 +55,28 @@ func ToAsciicast(ctx context.Context, session *bsr.Session, tmp storage.TempFile
...
-		if err != nil {
-			return nil, fmt.Errorf("%s: %w", op, err)
+			msgScanner, err := ch.OpenMessageScanner(ctx, bsr.Outbound)
+			if err != nil {
+				return nil, fmt.Errorf("%s: %w", op, err)
+			}
+
+			return sshChannelToAsciicast(ctx, reqScanner, msgScanner, tmp, options...)
+		default:
+			if chs.SessionProgram == "" {
```
[the remark this answers is not in the document]
Comment (elimt) [reply internal/bsr/convert/convert.go]: Yeah, I agree. That looks cleaner. Moved it as case statement

File: internal/bsr/convert/convert.go [hunk 2]
```diff
@@ -55,20 +55,27 @@ func ToAsciicast(ctx context.Context, session *bsr.Session, tmp storage.TempFile
 			return nil, fmt.Errorf("%s: %w", op, err)
 		}
 
-		// TODO sanity checks before getting the data files:
-		// - check connection summary to see if there was an exec or shell request
+		chs := ch.Summary.(*ssh.ChannelSummary)
```
Comment (tmessi) [inline internal/bsr/convert/convert.go:58]: This should either check the type assertion, i.e:

```golang
chs, ok := ch.Summary.(*ssh.ChannelSummary)
if !ok {
  // return error, probably indicating some like an unexpected or malformed bsr.
}
```

```go
switch chs := ch.Summary.(type) {
case *ssh.ChannelSummary:
  switch chs.SessionProgram {
  // rest of switch
default:
  // return an error, probably indicating some like an unexpected or malformed bsr.
```

Comment (tmessi) [reply internal/bsr/convert/convert.go:58]: Also, can you add a test case of a BSR that is malformed in this way?

File: internal/bsr/convert/convert_test.go [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+		enc, err := bsr.NewChunkEncoder(ctx, w, bsr.NoCompression, bsr.NoEncryption)
+		require.NoError(t, err)
+
+		for _, c := range chunks {
+			_, err := enc.Encode(ctx, c)
+			require.NoError(t, err)
+		}
+	}
+
+	cases := []struct {
```
Comment (tmessi) [inline internal/bsr/convert/convert_test.go:44]: Can you include test cases for all of the [valid ssh.SessionPrograms](https://github.com/hashicorp/boundary/blob/456c5dbb4d6c64886abcacb425b7faf5a39de673/internal/bsr/ssh/types.go#L39-L48). It looks like this already has some for `ssh.Exec`, `ssh.Shell` and `ssh.Sumsystem`, so can you add ones for `ssh.None` and `ssh.NotApplicable`?

File: internal/bsr/convert/convert.go [hunk 4]
```diff
@@ -55,20 +55,27 @@ func ToAsciicast(ctx context.Context, session *bsr.Session, tmp storage.TempFile
...
+				return nil, fmt.Errorf("%s: %w", op, err)
+			}
 
-		msgScanner, err := ch.OpenMessageScanner(ctx, bsr.Outbound)
-		if err != nil {
-			return nil, fmt.Errorf("%s: %w", op, err)
+			msgScanner, err := ch.OpenMessageScanner(ctx, bsr.Outbound)
+			if err != nil {
+				return nil, fmt.Errorf("%s: %w", op, err)
+			}
```
Comment (tmessi) [inline internal/bsr/convert/convert.go:70]: Can you add some tests that cover the error cases for these?

**Review by tmessi — CHANGES_REQUESTED** [review]
Looking pretty good.

**Review by ddebko — APPROVED** [review]
This looks good to me.

**Review by tmessi — APPROVED** [review]

**Review by irenarindos — APPROVED** [review]

**Pull request merged**